src/tests/end2end/SamplerTests.cpp

```diff
@@ -0,0 +1,195 @@
+
+        float verts[] = {
+            -2, -2, 0, 0,
+            -2,  2, 0, 0,
+             2, -2, 0, 0,
+            -2,  2, 0, 0,
+             2, -2, 0, 0,
+             2,  2, 0, 0,
+        };
+
+        mQuad = utils::CreateFrozenBufferFromData(device, verts, sizeof(verts), nxt::BufferUsageBit::Vertex);
+    }
+
+    void TestAddressModes(AddressModeTestCase u, AddressModeTestCase v, AddressModeTestCase w) {
+
+        nxt::Sampler sampler = device.CreateSamplerBuilder()
+            .SetAddressMode(u.mMode, v.mMode, w.mMode)
+            .GetResult();
+
+        auto bindGroup = device.CreateBindGroupBuilder()
+            .SetLayout(mBindGroupLayout)
+            .SetUsage(nxt::BindGroupUsage::Frozen)
+            .SetSamplers(0, 1, &sampler)
+            .SetTextureViews(1, 1, &mTextureView)
+            .GetResult();
+
+        uint32_t zeroOffset = 0;
+        nxt::CommandBuffer commands = device.CreateCommandBufferBuilder()
+            .BeginRenderPass(mFB.renderPass, mFB.framebuffer)
+            .BeginRenderSubpass()
+            .SetRenderPipeline(mPipeline)
+            .SetVertexBuffers(0, 1, &mQuad, &zeroOffset)
+            .SetBindGroup(0, bindGroup)
+            .DrawArrays(6, 1, 0, 0)
+            .EndRenderSubpass()
+            .EndRenderPass()
+            .GetResult();
+
+        queue.Submit(1, &commands);
+
+        RGBA8 expectedU2(u.mExpected2, u.mExpected2, u.mExpected2, 255);
+        RGBA8 expectedU3(u.mExpected3, u.mExpected3, u.mExpected3, 255);
+        RGBA8 expectedV2(v.mExpected2, v.mExpected2, v.mExpected2, 255);
+        RGBA8 expectedV3(v.mExpected3, v.mExpected3, v.mExpected3, 255);
+        RGBA8 black(0, 0, 0, 255);
+        RGBA8 white(255, 255, 255, 255);
+        EXPECT_PIXEL_RGBA8_EQ(black, mFB.color, 0, 0);
+        EXPECT_PIXEL_RGBA8_EQ(white, mFB.color, 0, 1);
+        EXPECT_PIXEL_RGBA8_EQ(white, mFB.color, 1, 0);
+        EXPECT_PIXEL_RGBA8_EQ(black, mFB.color, 1, 1);
+        EXPECT_PIXEL_RGBA8_EQ(expectedU2, mFB.color, 2, 0);
+        EXPECT_PIXEL_RGBA8_EQ(expectedU3, mFB.color, 3, 0);
+        EXPECT_PIXEL_RGBA8_EQ(expectedV2, mFB.color, 0, 2);
+        EXPECT_PIXEL_RGBA8_EQ(expectedV3, mFB.color, 0, 3);
+        // TODO: add tests for W address mode, once NXT supports 3D textures
+    }
+
+    // Test drawing a rect with a checkerboard texture with different address modes.
+    void DoAddressModeTest() {
+        for (auto u : addressModes) {
+            for (auto v : addressModes) {
+                for (auto w : addressModes) {
+                    TestAddressModes(u, v, w);
+                }
+            }
+        }
+    }
+
+    utils::BasicFramebuffer mFB;
+    nxt::BindGroupLayout mBindGroupLayout;
+    nxt::RenderPipeline mPipeline;
+    nxt::TextureView mTextureView;
+    nxt::Buffer mQuad;
+};
+
+TEST_P(SamplerTest, AddressMode) {
+    DoAddressModeTest();
+}
+
+NXT_INSTANTIATE_TEST(SamplerTest, D3D12Backend, MetalBackend, OpenGLBackend, VulkanBackend)
```

Comment:
This won't work on the Vulkan backend yet, can you remove it? Also did you check it work on D3D12, Metal and GL?

Reply:
Removed.

Tested on all backends; passes.